src/tools/pkcs11-tool.c

```diff
@@ -4486,7 +4496,52 @@ static int find_object(CK_SESSION_HANDLE sess, CK_OBJECT_CLASS cls,
 	if (rv != CKR_OK)
 		p11_fatal("C_FindObjects", rv);
 
-done:	if (count == 0)
+done:
+	if (count == 0)
+		*ret = CK_INVALID_HANDLE;
+	p11->C_FindObjectsFinal(sess);
+
+	return count;
+}
+
+static int find_object_by_label(CK_SESSION_HANDLE sess, CK_OBJECT_CLASS cls,
+		CK_OBJECT_HANDLE_PTR ret,
+		const char *label, int obj_index)
+{
+	CK_ATTRIBUTE attrs[2];
+	unsigned int nattrs = 0;
+	CK_ULONG count;
+	CK_RV rv;
+	int i;
+
+	attrs[0].type = CKA_CLASS;
+	attrs[0].pValue = &cls;
+	attrs[0].ulValueLen = sizeof(cls);
+	nattrs++;
+	if (label) {
+		attrs[nattrs].type = CKA_LABEL;
+		attrs[nattrs].pValue = (void *) label;
+		attrs[nattrs].ulValueLen = strlen(label);
+		nattrs++;
+	}
```

Comment:
I think it is ok now, but after having a second look, I would probably like to see the original `find_object()` modified to support searching also by the label just by adding the label argument and this block instead of duplicating all the rest of this function.

Reply:
For minimal changes, I split it into find_object_id_or_label and find_object functions.